File: examples/04_struct_introspection.ts
function main(): void {

interface Point {
  x: number;
  y: number;
}

  function createPoint(x: number, y: number): Point {
    return {
      x: x,
      y: y
    };
  }

  const POINT_SIZE: number = 2;

interface Color {
  r: number;
  g: number;
  b: number;
}

  function createColor(r: number, g: number, b: number): Color {
    return {
      r: r,
      g: g,
      b: b
    };
  }

  const COLOR_SIZE: number = 3;

  const POINT_SIZE: number = 16;

  const COLOR_SIZE: number = 24;

  const POINT_FIELDS: number = 2;

  const COLOR_FIELDS: number = 3;

  const POINT_HAS_X: boolean = true;

  const POINT_HAS_Z: boolean = false;

  const POINT_METHODS: number = 0;

  const COLOR_METHODS: number = 0;
  console.log("=== Struct Introspection ===");
  console.log(`Point size: ${POINT_SIZE} bytes`);
  console.log(`Color size: ${COLOR_SIZE} bytes`);
  console.log(`Point fields: ${POINT_SIZE}`);
  console.log(`Color fields: ${COLOR_SIZE}`);
  console.log(`Point has x: ${true}`);
  console.log(`Point has z: ${false}`);
  console.log(`Point methods: ${0}`);
  console.log(`Color methods: ${0}`);
  console.log("\n✓ Introspection completed!");
  console.log("\n=== Transpilation Demo ===");

  const POINT_SIZE_CONST: number = 16;

  const COLOR_SIZE_CONST: number = 24;

  const TOTAL_SIZE: number = 40;
  let origin = createPoint(0.0, 0.0);
  let red = createColor(255, 0, 0);
  console.log("Transpilation target sizes:");
  console.log(`  Point: ${POINT_SIZE_CONST} bytes (const)`);
  console.log(`  Color: ${COLOR_SIZE_CONST} bytes (const)`);
  console.log(`  Combined: ${TOTAL_SIZE} bytes`);
  console.log("Runtime instances:");
  console.log(`  Origin: (${origin.x}, ${origin.y})`);
  console.log(`  Red: rgb(${red.r}, ${red.g}, ${red.b})`);
  console.log("\n✓ Introspection enables external code generation!");
}

main();

File: examples/04_struct_introspection.rs
pub fn main() -> () {
    pub struct Point {
        pub x: f64,
        pub y: f64,
    }
    pub struct Color {
        pub r: u8,
        pub g: u8,
        pub b: u8,
    }
    pub const POINT_SIZE: usize = 16;
    pub const COLOR_SIZE: usize = 24;
    pub const POINT_FIELDS: usize = 2;
    pub const COLOR_FIELDS: usize = 3;
    pub const POINT_HAS_X: bool = true;
    pub const POINT_HAS_Z: bool = false;
    pub const POINT_METHODS: usize = 0;
    pub const COLOR_METHODS: usize = 0;
    println!("=== Struct Introspection ===");
    println!("Point size: {} bytes", intrinsic_size_of(Point));
    println!("Color size: {} bytes", intrinsic_size_of(Color));
    println!("Point fields: {}", intrinsic_field_count(Point));
    println!("Color fields: {}", intrinsic_field_count(Color));
    println!("Point has x: {}", intrinsic_has_field(Point, "x"));
    println!("Point has z: {}", intrinsic_has_field(Point, "z"));
    println!("Point methods: {}", intrinsic_method_count(Point));
    println!("Color methods: {}", intrinsic_method_count(Color));
    println!("\n✓ Introspection completed!");
    println!("\n=== Transpilation Demo ===");
    pub const POINT_SIZE_CONST: usize = 16;
    pub const COLOR_SIZE_CONST: usize = 24;
    pub const TOTAL_SIZE: usize = 40;
    let origin = Point { x: 0, y: 0 };
    let red = Color { r: 255, g: 0, b: 0 };
    println!("Transpilation target sizes:");
    println!("  Point: {} bytes (const)", POINT_SIZE_CONST);
    println!("  Color: {} bytes (const)", COLOR_SIZE_CONST);
    println!("  Combined: {} bytes", TOTAL_SIZE);
    println!("Runtime instances:");
    println!("  Origin: ({}, {})", origin.x, origin.y);
    println!("  Red: rgb({}, {}, {})", red.r, red.g, red.b);
    println!("\n✓ Introspection enables external code generation!");
}
